### ISSUE TITLE
Low power support for RP2350

### ISSUE THREAD

**Issue by tannewt** [adafruit/circuitpython#9491]
The SDK doesn't have an official API for it yet. So, we can wait or DIY it.

**Comment by tannewt** [comment]
Reference code is linked to (and discussed) in this issue: https://github.com/raspberrypi/pico-examples/issues/530

**Comment by dmcomm** [comment]
I'm trying to run an RP2040 program that uses `sleep_memory` to pass data between resets and between boot.py and main program, but doesn't actually use any of the low-power features. Not sure if this is a common situation or not.

**Comment by bablokb** [comment]
@eightycc: since you self-assigned this, please have a look at my PR. It works fine on all of my systems. And in contrast to the discussion, it does not "change semantics" (except for a single academic border case which is not relevant for any practical real-life scenario).

It *does* change the documentation bringing it in line with reality. The docs currently claims things that the code does not stand up to and never will ("It does not shut down WiFi, BLE, or other communications, or ongoing activities such as audio playback").

**Comment by eightycc** [comment]
Thank you, @bablokb! I'll be looking closely at your work on #9965 and #10023.

**Comment by eightycc** [comment]
As I've been working through implementing the PR for this issue, I've put together these notes.

#### The Pico processors and their SDK define the following low-power states:

- SLEEP state - The SLEEP state is entered when both cores are asleep which is: (1) both cores have entered SLEEP by means of WFI/WFE, and (2) DMA is quiescent. Upon entering the SLEEP state, portions of the clock tree are gated OFF under control of CLOCKS SLEEP_ENx registers. In the SLEEP state, power dissipation is reduced for those clock boundaries that are gated OFF, while the logic within those boundaries that are gated OFF retains its state. SLEEP state is
exited when either core is awoken by an interrupt.

  Note: The SLEEPDEEP bit is an ARM core feature that controls whether WFI/WFE enters the SLEEP state or the DEEP SLEEP state. According to ARM, both states are identical for the internal operation of the processor core. Setting the SLEEPDEEP does assert a SLEEPDEEP signal from the core. Presumably, the Pico parts use this signal as an overall control for clock gating.

  CircuitPython uses the SLEEP state to implement its "light sleep" and for "deep sleep" cases where DORMANT state cannot be used.

- DORMANT state - The DORMANT state is entered by writing to memory mapped registers defined for that purpose. A DORMANT register exists in both the XOSC and ROSC register spaces, the oscillator currently in use owns the DORMANT register to be used. Unlike entry to SLEEP state, the driving code is responsible for ensuring that the second core and DMA are quiescent before triggering entry to the DORMANT state. In the DORMANT state, all clocks and oscillators are disabled, reducing dynamic power dissipation to zero. It is the responsibility of the driving software to turn off PLLs before entering the DORMANT state. DORMANT state is exited by a GPIO event, or an RTC/AON  interrupt.

  Whenever possible, CircuitPython uses the DORMANT state to implement its "deep sleep".

The SLEEP and DORMANT states are aimed at reducing dynamic power. Dynamic power is dissipated by the operation of logic while it is clocked. Additional controls are provided that allow the reduction of static power by powering down specific power domains.
- Memory Periphery Power Down - Powers down the analog circuitry used to access SRAM, but retains its contents. 
- Power Management - Defines a number of low-power states where various power boundaries may be switched off selectively.
- Flash Memory Power Down - Flash memories that implement the Power-down (xB9) command may be put into a power-down state for additional power savings beyond their usual standby power dissipation.

#### CircuitPython considerations:
1. CircuitPython uses an area of SRAM to implement sleep memory. Sleep memory retains its state through all defined sleep transitions, allowing an application to retain a small amount of state. Use of sleep memory precludes complete RP2350 SRAM power down. While it may be possible to isolate sleep memory in one of the two SRAM power domains, CircuitPython does not attempt to do so.
2. For the RP2040, deep sleep cannot use the DORMANT state when a timer alarm is requested. This is because the RTC clock source must be kept running from either the XOSC or ROSC.
3. For the RP2350, the AON timer has severely limited accuracy when it uses the LPOSC as its tick source. Since the accurate XOSC clock is part of the SWCORE power domain, powering down the SWCORE power domain is not possible if accuracy of better than +/- 20% for AON alarm triggered exit from one of the P1 power management states is desired. For the RP2350, CicruitPython enters the SLEEP state when timer alarm exit from "deep sleep" is requested. When timer alarm exit from deep sleep is not requested, CircuitPython reduces static power dissipation as far as possible while retaining SRAM contents by entering the P1.4 low-power state.
4. For the RP2350, the number of GPIO sources that can exit from a P1 power management state is limited to 4. If more than 4 GPIO wakeup alarms are requested, the implementation will fall back to using the DORMANT state.
5. CircuitPython always exits true deep sleep by means of a reset. The reset may be triggered automatically by power management exiting a P1 state or programmatically when exiting the DORMANT state.
6. Sleep clock gating is used by CircuitPython as follows:
   a. For light sleep, all non-essential clock domains are gated OFF. Essential clock domains are those that can generate interrupts to wake the core including timers, GPIO pads, and the USB peripheral.
   b. For RP2040, DORMANT state cannot be used for deep sleep with a timer alarm as noted above in (2). Sleep clock gating is used in this case to reduce dynamic power dissipation as much as possible by gating OFF all clock domains except the timer and the GPIO pads.
7. Additional static power savings may by be achieved through the use of memory periphery power down. CircuitPython does not presently use this feature.
8. Because DORMANT mode powers off all clocks, the RTC and AON timer are affected even if they are not used as an alarm source. For the RP2040, the RTC will stop running while in DORMANT state. For the RP2350, the AON timer will continue running from the LPOSC, but its accuracy will be affected while in the DORMANT state. 

#### Possible future enhancements:
1. For the RP2040 and RP2350 we use a generic implementation of "fake deep sleep" when USB or a wireless workflow is active. A specialized implementation of "fake deep sleep" could use SLEEP mode to achieve power saving similar to "light sleep".
2. For the RP2040 and RP2350, when a timer alarm is set, we use SLEEP state instead of DORMANT state (or possibly low-power state P1.4 for RP2350) for "deep sleep". As previously noted, this is done to allow the the RTC or AON timer to continue running from the XOSC, ensuring timer alarm accuracy. Enhancing CircuitPython's RTC module to allow definition of an external 1 kHz tick source would allow use of the DORMANT state for "deep sleep" with a timer alarm. This would have the added benefit of keeping the RTC or AON timer running in "deep sleep" when a timer alarm is not in use.
3. Exposing an additional accuracy control for "deep sleep" that would allow the user to choose between power savings and timer alarm accuracy.
4. Memory Periphery Power Down. Use of this feature with DORMANT state may provide additional unknown power savings.
5. Power Manager low-power states. Switching off all power domains in low-power state P1.7 would accrue significant power savings. This would mean complete loss of SRAM contents, making sleep memory unusable. There are also SDK and possible power manager problems to overcome to enter and exit this state successfully.
6. Flash memory power down could save power, depending on the flash part. Using it would require special code to run after reset that would take the part out of its power-down state.

**Comment by bablokb** [comment]
Some more notes:

- powman resets directly after wakeup, so you cannot set the wakeup cause
- On the RP2350, the AON-timer uses the powman-timer internally.
- as a user, I would want the choice between power-savings and timer accuracy. If I need an accurate timer, I use an external RTC and GPIO-wakeup anyhow. LPOSC accuracy can be tested and taken into account for. I have specimen with 1%, 8% and 25% differences. These are not constants and depend e.g. on temperature, but there are use-cases where that is good enough. As an example, I have sensors that need about 30s warmup time. There is no need to accurately "light-sleep" the RP2350 at high currents during this time.
- I have experimental code that uses powman-states for deep-sleep. Enter and exit works fine for timers, but I have problems with the reliability of GPIOs. With a timer, I have something like 600µA, with a GPIO I have 150µA. This is all very preliminary.

**Comment by eightycc** [comment]
> powman resets directly after wakeup, so you cannot set the wakeup cause

The wake-up cause can be determined by examining POWMAN register LAST_SWCORE_PWRUP after reset. My implementation leaves a special value in SCRATCH0 to indicate this case, although CHIP_RESET with HAD_SWCORE_PD set likely tells the same story.

> On the RP2350, the AON-timer uses the powman-timer internally.

As far as I can tell, the AON timer _is_ the POWMAN timer.

> I have experimental code that uses powman-states for deep-sleep.

Would you be willing to share your code? I'm doing quite a bit of stumbling around in the dark with P1 states. Which P1 state are you entering for those impressive power numbers?

The LPOSC is crazy inaccurate and as you note extremely temperature sensitive. I've seen noted on the RPi discussion board that any attempt to calibrate it for use in low-power states is problematic due to the change in die temperature between a P0 and a P1 state.

**Comment by bablokb** [comment]
> As far as I can tell, the AON timer _is_ the POWMAN timer.

As far as I understand the SDK, this is only true for the RP2350.

> Would you be willing to share your code? I'm doing quite a bit of stumbling around in the dark with P1 states. Which P1 state are you entering for those impressive power numbers?

It is still too experimental but I will push my powman-tree once it is cleaned up and a bit more tested. But I mainly follow the examples from https://github.com/peterharperuk/pico-examples/tree/powman/powman. This code uses 

```
// Power states
    powman_power_state P1_7 = POWMAN_POWER_STATE_NONE;

    powman_power_state P0_3 = POWMAN_POWER_STATE_NONE;
    P0_3 = powman_power_state_with_domain_on(P0_3, POWMAN_POWER_DOMAIN_SWITCHED_CORE);
    P0_3 = powman_power_state_with_domain_on(P0_3, POWMAN_POWER_DOMAIN_XIP_CACHE);

    off_state = P1_7;
    on_state = P0_3;
```

> ... is problematic due to the change in die temperature between a P0 and a P1 state.

I am not that pessimistic, since once I am in P0 state, I am in P0 state. In all of my tests, the inaccuracy was rather deterministic for a given specimen. Nevertheless, I would not vote for the LPOSC as the default, but I would strongly support to have the option to use it. (if it is worth it - I haven't measured the dormant-state using XOSC yet).

I am currently fighting with GPIO wakeup. It is really strange, only about one out of three times the pico will enter the off-state. It seems like there is a glitch that triggers wake up already a fraction of a second after the device powered down.  And I have my GPIO hardwired to 3V3 so I cannot really understand why.

**Comment by eightycc** [comment]
> I am currently fighting with GPIO wakeup. It is really strange, only about one out of three times the pico will enter the off-state. It seems like there is a glitch that triggers wake up already a fraction of a second after the device powered down. And I have my GPIO hardwired to 3V3 so I cannot really understand why.

Some of the power manager registers do not reset on a power manager triggered reset. Calling `powman_disable_all_wakeups()` before setting up new alarm conditions for entry to a low-power state will ensure you don't exit again for a previously triggered alarm condition. It's not in the example code, which is why I mention it.

**Comment by bablokb** [comment]
It is really time that I push my working tree.. I already have these calls in my code:

```
        powman_disable_all_wakeups();
        powman_timer_disable_alarm();
        powman_clear_alarm();
```

The strange thing is that the code works sometimes. My test program just blinks the LED a few times and then tries to power-down. If I just let it run, it takes a random number of boot-cycles until the power-down is successful.

**Comment by bablokb** [comment]
I pushed my working tree: https://github.com/bablokb/circuitpython/tree/rp2xxx_powman

Note that this is based on 9.2.8.

I am also adding an image from ppk2 that demonstrates the behavior I described above. In this test, the system needed four attempts to enter the power-off state, but that count is random.

![Image](https://github.com/user-attachments/assets/bad5a6c3-acad-4b98-bbda-b3388cb84855)

**Comment by eightycc** [comment]
@bablokb You've probably already figured this out, but I'll mention it just in case: The SWD interacts with the operation of the RP2350's power sequencer and power manager in some undocumented ways. With `debugprobe` attached and `openocd` active, power manager registers `CHIP_RESET` and `LAST_SWCORE_PWRUP` (and likely some others) are affected after restart from an SWCORE power-down. I've found that it's safe to have `debugprobe` physically attached, but with `openocd` running things get weird.

**Comment by eightycc** [comment]
See https://github.com/raspberrypi/pico-sdk/pull/2516 for fixes to Pico SDK power manager support.

**Comment by bablokb** [comment]
`LAST_SWCORE_PWRUP` does not seem reliable even without any debugprobe attached. It always reads `2` (which is pwrup1 according to the docs), but I am only testing with a single GPIO and I am using pwrup0 which should turn up as `1`.

Also, after a chip reset I see a `1` instead of the documented `0`.

**Comment by dwrrehman** [comment]
> `LAST_SWCORE_PWRUP` does not seem reliable even without any debugprobe attached. It always reads `2` (which is pwrup1 according to the docs), but I am only testing with a single GPIO and I am using pwrup0 which should turn up as `1`.
> 
> Also, after a chip reset I see a `1` instead of the documented `0`.

Hi, I am currently implementing the P1.x low power modes from scratch in RISC-V assembly (without a debugger at all), and am only using the RP2350 datasheet to write this program. I came across this confusion at first as well, when the RP2350 datasheet says the following in the LAST_SWCORE_PWRUP (POWMAN) register in chapter 6:

> Indicates which pwrup source triggered the last switched-core power up:
> 0 = chip reset, for the source of the last reset see POWMAN_CHIP_RESET 
> 1 = pwrup0
> 2 = pwrup1
> 3 = pwrup2
> 4 = pwrup3
> 5 = coresight_pwrup
> 6 = alarm_pwrup

what it really means is:

> Indicates which pwrup source triggered the last switched-core power up:
> bit 0 = if 1, then indicates chip reset, for the source of the last reset see POWMAN_CHIP_RESET 
> bit 1 = if 1, then indicates pwrup0
> bit 2 = if 1, then indicates pwrup1
> bit 3 = if 1, then indicates pwrup2
> bit 4 = if 1, then indicates pwrup3
> bit 5 = if 1, then indicates coresight_pwrup
> bit 6 = if 1, then indicates alarm_pwrup

This explains the values you were seeing, and also can be deduced by the fact that the LAST_SWCORE_PWRUP register contains 7 bits, (bits 0 through 6). Taking this into account in my programs, yields this register always having the correct value according to the datasheet, i've never seen it do anything unreliable or odd so far. 


As of today, I have gotten low-power mode LED-toggling programs working on the Pico 2 and 2W, using low-power modes P1.4 / P1.6 (SRAM retention seems to be required for any sort of stateful low-power application on the RP2350), and using both types of possible wake-ups on the RP2350: 

- using GPIO wake-ups (via PWRUP0 on GPIO pin 15), and 
- using ALARM_TIMER (which is synonymous with the AON timer, as well as the POWMAN timer, on the RP2350). 

From my testing with a current-measuring multimeter in series with 5v being applied to the VSYS pin, I have gotten both GPIO-based and ALARM-based programs sleeping at currents of less than 210uA on the Pico 2W boards, and less than 240uA on the Pico 2 boards. 

(The discrepancy between the Pico 2 and Pico 2W here is due to the schematic on the Pico 2 containing some current-consuming resistors in the analog input circuitry, and a lot of the other current here is from the regulator/other circuitry besides the RP2350 itself, which should be able to sleep down to around 19uA / 68uA / 120uA (depending on how much SRAM domains you have enabled) according the RP2350 datasheet.)

If it helps save you some trouble, I did notice I had to do something in my program that was quite non-obvious in order to allow multiple time-consecutive GPIO wakeups, and the datasheet doesn't spell this out in great detail. To re-enable a GPIO wakeup on a given pin after a low-power mode GPIO wakeup, I had to write to PWRUP0 twice: once with {STATUS=0b1,ENABLE=0b0}, and then again with {STATUS=0b1,ENABLE=0b1}. I deduced I had to do this, due to the datasheet saying the following under the ENABLE field of the PWRUP0 register:

> ENABLE: Set to 1 to enable the wakeup source. Set to 0 to disable the wakeup source and clear a pending wakeup event.
If using edge detect a latched edge needs to be cleared by writing 1 to the status register also.

....which, from my interpretation sounds like you genuinely need to clear a pending wake up event first, separately, and then can enable the new GPIO wake up again after that past event has been cleared. 


Anyways, hope this helps! The Pico SDK doesn't seem totally exhaustive on this subject of low power modes, so i'd be happy to share my RISC-V assembly code / pseudocode for my register writes if you ever get stuck further down the line. :)

**Comment by bablokb** [comment]
Yes, I totally misunderstood the docs regarding `LAST_SWCORE_PWRUP`.

My work on this has stalled recently due to other things (and because I was out of ideas), but I will revisit it and check if I do the correct things you are pointing out. From my guts feeling, it still does not explain the random behavior I can see: instant wake up after going to low-power. This happens a few ms after low-power is reached. Ok, this sounds like a wake up from an alarm that wasn't cleared, but the random behavior also happens after POR.

Regarding sleep-current: I can see 154µA@5V on a Pico2:

<img width="1025" height="630" alt="Image" src="https://github.com/user-attachments/assets/6d4113ae-9120-4ff7-92e9-8290f369f73d" />

So that's in line with your results.

**Comment by adeldigital** [comment]
@bablokb Thanks for sharing the PPK2 screenshots, they are very useful. Could you confirm if I understood correctly, you got 154µA@5V during sleep with SRAM retention? Is that in the dormant mode?

**Comment by bablokb** [comment]
No, this is neither dormant mode, nor with SRAM retention. These are figures from my powman experiments. On wake up, the system resets so nothing in SRAM is retained.

In dormant-mode with state retention (aka "light-sleep") I am down to 1.9mA. See #10023 for details and a lengthy discussion about why it seems to be better to have no alarm-implementation for the RP2350 rather than to have at least a working implementation.

**Comment by adeldigital** [comment]
@bablokb Oh, okay. So 1.9mA is the absolute lowest power sleep state the RP2350 can reach with SRAM retention? Shoot, I need my system way lower than that. Looks like I need to look for a different microcontroller. Thanks

**Comment by bablokb** [comment]
Yes, 1.9mA is the current state of art. Theoretically, powman can also keep SRAM state but that would only help if you can continue where you left of. And I have not seen any working code that does this.

For real low power usage you either need a mcu that is optimized for that purpose, or use some external circuitry. But that would also not keep state.

**Comment by dwrrehman** [comment]
> [@bablokb](https://github.com/bablokb) Oh, okay. So 1.9mA is the absolute lowest power sleep state the RP2350 can reach with SRAM retention? Shoot, I need my system way lower than that. Looks like I need to look for a different microcontroller. Thanks

Actually ~188uA is the lowest the RP2350 can go with 256kB of SRAM retained (SRAM power domain 0 or domain 1 (not both) specifically), specifically using low power mode "P1.6" (or "P1.5") as i mentioned above. Like Bablokb mentioned, this does involve a CPU reset (as it is being power-gated in P1.x modes), however SRAM contents are not reset on a CPU reset of this kind, IIRC.

(note, this is using low power mode P1.6 (or P1.5), ie not the RP2350's "Sleep" mode, nor it's "Dormant" mode. Sleep and Dormant modes are distinct modes with somewhat low power consumption. but these do not power-gate anything in the chip, they simply gate the clocks to the cpu, or other simple power saving methods like that).

**Comment by dwrrehman** [comment]
> Yes, 1.9mA is the current state of art. Theoretically, powman can also keep SRAM state but that would only help if you can continue where you left of. And I have not seen any working code that does this.
> 
> For real low power usage you either need a mcu that is optimized for that purpose, or use some external circuitry. But that would also not keep state.

Interestingly, one RISC-V assembly program I have been recently experimenting with, is the possibility of allowing the system to simply resume where it was prior to the sleep call, however, still sleep at 188uA. 

The technique is rather clever: instead of just putting the system into sleep mode using low power mode P1.6 or P1.5, you first store the state of all 31 registers into particular predefined SRAM locations, and then once those are stored, you configure the system to go into sleep using P1.5 or P1.6. Then, on CPU reset/boot, you check the LAST_SWCORE_PWRUP register, and if it indicates a GPIO or ALARM-based low power wake up event, then you load the values from those predefined SRAM locations into the CPU registers, and then branch to the stored PC value + 4 (using a JALR, or some register branch). 

This allows you to sleep at 188uA current levels, while still allowing you to pick up execution exactly where you left off, right after the WFI instruction (which put the CPU into P1.5/P1.6). For a RISC-V assembly program, implementing this logic is fairly trivial, i just haven't tested it yet. Unfortunately, however, I get the feeling that an optimization/mechanism such as this could only really be done at the assembly level, or would be slightly difficult in high level languages, unless I am mistaken. Perhaps someone else might know if this would even be feasible to try to implement! :)

**Comment by bablokb** [comment]
There has been some work done on this, e.g. here: https://github.com/mamba2410/rp2350-powman-sleep, see also
https://forums.raspberrypi.com/viewtopic.php?t=384176#p2297511

But this does not sound as if it were really usable ("I managed to get this working, mostly!").

Anyhow, before investigating this, I will first try to make the system enter sleep correctly for gpio-wakeups.

**Comment by dwrrehman** [comment]
> There has been some work done on this, e.g. here: https://github.com/mamba2410/rp2350-powman-sleep, see also https://forums.raspberrypi.com/viewtopic.php?t=384176#p2297511
> 
> But this does not sound as if it were really usable ("I managed to get this working, mostly!").
> 
> Anyhow, before investigating this, I will first try to make the system enter sleep correctly for gpio-wakeups.

Thanks for linking the page! This is fascinating work, I was not aware this already existed! Hopefully eventually a polished solution/API for a low power mechanism like this comes together at some point. I know it sure would be quite useful in many battery-powered RP2350 applications!

Best of luck in your gpio-wakeup testing! feel free to reach out for any help or questions, i'd be happy to help in any way i can.

**Comment by bablokb** [comment]
> Best of luck in your gpio-wakeup testing! feel free to reach out for any help or questions, i'd be happy to help in any way i can.

Thanks! I think you just triggered a change of my project priorities 😅️

**Comment by adeldigital** [comment]
@dwrrehman 188 µA sounds much better. In my case I don’t need the CPU to continue where it left off. I just need to keep an API key in SRAM retained. If I don’t need the CPU to resume its previous state, could allowing it to start from reset (or something along those lines) reduce the power consumption even further?

**Comment by bablokb** [comment]
You can store it in nvram and then you could even turn of power.

Be aware that all of this discussed here is very experimental, we don't even have official support for the alarm-module for the RP2350. Your out-of-the-box experience will be better with an ESP32-S3 or a Nordic mcu.

**Comment by dwrrehman** [comment]
> [@dwrrehman](https://github.com/dwrrehman) 188 µA sounds much better. In my case I don’t need the CPU to continue where it left off. I just need to keep an API key in SRAM retained. If I don’t need the CPU to resume its previous state, could allowing it to start from reset (or something along those lines) reduce the power consumption even further?

It seems allowing the CPU to start from reset will give you equivalent low power consumption, 188uA, (low power mode P1.5/P1.6, where 1 bank of SRAM retention is enabled) as adding this (currently hypothetical) mechanism to let the CPU pick up where it left off in execution, from my understanding. 

However, as bablokb mentioned, for your use case its entirely possible you could be able to store the API key in some form of non-volatile memory (NVRAM), such as an FRAM chip, MRAM chip, BBSRAM, etc. Technically using the Flash could work, however only if this data does not need to be written often. Once the data is written to some non-volatile memory, you can simply unpower the RP2350 entirely using an external circuit instead of using the sleep modes.

**Comment by Skorpionm** [comment]
Thanks for the tips, I FINALLY GOT IT WORKING!!! The program continues to work with the next function where it went into sleep mode with full SRAM saving.
But there's some magic! The do_sleep and exit_sleep functions are placed in memory, and exit_sleep is modified with __attribute__((naked)), so that it can make an unconditional jump to another part of the program without breaking the stack.
do_sleep remembers the return address and the stack size allocated for it
app_loop optimaze O0

```
app_loop{
 ...
asm volatile (
        "mov %0, sp"
        : "=r" (stack_pointer_do_sleep)
    );
do_sleep(stack_pointer_do_sleep); 
...
}

bool  __noinline __not_in_flash_func(do_sleep)(uintptr_t start_stack_pointer) {
 lr = (uintptr_t)__builtin_extract_return_addr (__builtin_return_address (0));
....
    powman_hw->boot[0] = 0xb007c0d3; // magic_number
    powman_hw->boot[1] = boot_addr ^ -0xb007c0d3; // (-magic_number) ^ boot_addr
    powman_hw->boot[2] = start_stack_pointer; // stack pointer
    powman_hw->boot[3] = exit_sleep_addr; // boot_addr
.....
while (true) __wfi();
}

void __attribute__((naked)) __noinline __not_in_flash_func(exit_sleep)() {
 .....
// Jump back to where we were before sleep
    asm volatile (
        "mov pc, %0"
        :
        : "r" (lr)
    );
}
```
and this magic has been working for 3 days now without freezing or resetting, 10ms of sleep, 10ms of work
<img width="2092" height="1482" alt="Image" src="https://github.com/user-attachments/assets/7415959b-89b7-4b2f-a349-1442d06e8125" />

<img width="3840" height="2100" alt="Image" src="https://github.com/user-attachments/assets/d3c19e9b-05c0-4d23-8eb9-d299ac0aa931" />

My implementation provides for the possibility of exiting sleep also by interruption. When exiting in this mode, it is also necessary to reset all states when exiting sleep, otherwise the code will not be able to wake up again.

**Comment by Skorpionm** [comment]
Freetos still needs to learn this technique.

**Comment by bablokb** [comment]
@Skorpionm Thanks for sharing! Do you also have the complete program somewhere?

**Comment by Skorpionm** [comment]
No!

**Comment by bablokb** [comment]
@dwrrehman 

The status-bit makes me scratch my head... As you wrote

> To re-enable a GPIO wakeup on a given pin after a low-power mode GPIO wakeup, I had to write to PWRUP0 twice: once with {STATUS=0b1,ENABLE=0b0}, and then again with {STATUS=0b1,ENABLE=0b1}. I deduced I had to do this, due to the datasheet saying the following under the ENABLE field of the PWRUP0 register:
> 
> > ENABLE: Set to 1 to enable the wakeup source. Set to 0 to disable the wakeup source and clear a pending wakeup event.
> > If using edge detect a latched edge needs to be cleared by writing 1 to the status register also.
> 
> ....which, from my interpretation sounds like you genuinely need to clear a pending wake up event first, separately, and then can enable the new GPIO wake up again after that past event has been cleared.

This is in line with the datasheet and the file `rp2350/hardware_regs/include/hardware/regs/powman.h`:

```
// -----------------------------------------------------------------------------
// Field       : POWMAN_PWRUP0_STATUS
// Description : Status of gpio wakeup. Write to 1 to clear a latched edge
//               detect.
#define POWMAN_PWRUP0_STATUS_RESET  _u(0x0)
#define POWMAN_PWRUP0_STATUS_BITS   _u(0x00000200)
#define POWMAN_PWRUP0_STATUS_MSB    _u(9)
#define POWMAN_PWRUP0_STATUS_LSB    _u(9)
#define POWMAN_PWRUP0_STATUS_ACCESS "WC"
```

Now the first `#define` already contradicts the "write to 1". And the sdk-function `powman_enable_gpio_wakeup()` in fact clears the status-bit by writing a 0 (followed by a second write that sets the enable bit). Also, when I print the pwrup0 register after POR, the status-bit is zero. After the wakeup is enabled, it is still zero and then after the wakeup it is 1.

Anyhow, regardless of what I do with the status bit it actually does not make any difference for me. I still encounter these random instantaneous wakeups directly after power down and that is independent of edge or level detect or clearing/setting the status-bit.

**Comment by dwrrehman** [comment]
Hi, @bablokb,

Thanks for the question! :) I'll see if I can illuminate whats happening a bit more, if you'd like!

Below I've included the implementation of `powman_enable_gpio_wakeup` for our reference.

```c
254 | void powman_enable_gpio_wakeup(uint gpio_wakeup_num, uint32_t gpio, bool edge, bool high) {
255 |     invalid_params_if(HARDWARE_POWMAN, gpio_wakeup_num >= count_of(powman_hw->pwrup));
256 | 
257 |     // Need to make sure pad is input enabled
258 |     gpio_set_input_enabled(gpio, true);
259 | 
260 |     // Set up gpio hardware for what we want
261 |     uint32_t pwrup = (edge ? POWMAN_PWRUP0_MODE_VALUE_EDGE : POWMAN_PWRUP0_MODE_VALUE_LEVEL) << POWMAN_PWRUP0_MODE_LSB;
262 |     pwrup |= (high ? POWMAN_PWRUP0_DIRECTION_BITS : 0);
263 |     pwrup |= gpio << POWMAN_PWRUP0_SOURCE_LSB;
264 |     powman_write(&powman_hw->pwrup[gpio_wakeup_num], pwrup);
265 | 
266 |     // Clear the status bit in case an edge is already latched
267 |     powman_clear_bits(&powman_hw->pwrup[gpio_wakeup_num], POWMAN_PWRUP0_STATUS_BITS);
268 | 
269 |     // Important to enable it separately to allow the gpio to change
270 |     powman_set_bits(&powman_hw->pwrup[gpio_wakeup_num], POWMAN_PWRUP0_ENABLE_BITS);
271 | }
```

Of important note here is that on line `267` we use the value `POWMAN_PWRUP0_STATUS_BITS`, which has the value `0x00000200`:

```c
#define POWMAN_PWRUP0_STATUS_BITS   _u(0x00000200)
```

which equates to having a `1` bit set at bit position `9`, which the datasheet says is the `STATUS` bit/field. Digging deeper, we see that `powman_clear_bits` has the following definition:

```c
124 | static inline void powman_clear_bits(volatile uint32_t *reg, uint32_t bits) {
125 |     invalid_params_if(HARDWARE_POWMAN, bits >> 16);
126 |     hw_clear_bits(reg, POWMAN_PASSWORD_BITS | bits);
127 | }
```
...And diving even further, to see what actual underlying C code will be used, we see:

```c

112 | #define hw_clear_alias_untyped(addr) ((void *)(REG_ALIAS_CLR_BITS + hw_alias_check_addr(addr)))
...	...
145 | __force_inline static void hw_clear_bits(io_rw_32 *addr, uint32_t mask) {
146 |     *(io_rw_32 *) hw_clear_alias_untyped((volatile void *) addr) = mask;
147 | }
```

...Which uses the following definition, found in `addressmap.h`:

```c
21 | #define REG_ALIAS_CLR_BITS (_u(0x3) << _u(12))
```

it thus seems that `REG_ALIAS_CLR_BITS` effectively has the value `0x3000`, and we are effectively OR'ing it with the `PWRUP0` register's address, to obtain the actual address we write to, and the actual value we write is the bits _set_ to `1`. Looks like what is happening here, is that the SDK opts to (confusingly) use the register address aliases, which are outlined in the RP2350 datasheet in section `2.1.3.`, titled `Atomic register access`. if `PWRUP0` has address `A`, then we are writing to:

```c
*(A | 0x3000) = 0x00000200;
```

However, this bit we are writing to here, is of the register type "`WC`", meaning "Cleared-when-writing-a-1". This is because "`WC`" appears next to the bitfield name in `PWRUP0`'s register documentation, and `Appendix A: Register field types` of the RP2350 datasheet defines `WC` as the following:

```
WC: 
  - Write-Clear
  - Writing a 1 to a bit in a WC field will write that bit to 0
  - Writing a 0 to a bit in a WC field does nothing
  - Read operation returns the register value
```

And although I can't find this exact page which talks about the combination of these two semantics, i think its clear (pardon the pun) that using the "Clear-on-write" register address aliases on a `WC` bitfield yields the same (somewhat redundant) semantics, of writes of a 1 cause the field to clear to be a 0. This register bitfield's value is always set to `1` by hardware (software can only clear it), and given that fact, reads work as you would expect, I think. Connecting this to what I mentioned earlier, it seems that you can only enable a GPIO wakeup once this bit is clear (ie would read as a 0, if read). 

Finally, I have been thinking about what might be the cause for your spurious wakeup events being triggered directly after going to sleep, and I cannot quite think of what might be triggering that. On the bright side, i have a hunch that it might be due to interactions of some other piece of code with the POWMAN system, for example, generated C code / CPU instructions involved in controlling something else on the RP2350 other than POWMAN's low power mode interface itself, and this seperate system is somehow sending events, IRQs, or something else which is causing the CPU to wake up. 

I hypothesize this mainly because when I use my RISC-V assembler and manually program the direct register writes I want to happen in sequence, and flash the UF2 generated by the assembler to my Pico 2 boards, the program works flawlessly with no spurious wakeups whatsoever (at least, none that i have been able to detect, anyways). This leads me to conclude that if there is no other conflicting interactions happening on the system, and one has the right sequence of register writes to the POWMAN subsystem, low power modes, and GPIO wakeups / ALARM wakeups work flawlessly. It's possible I could be wrong about this hypothesis, however. 

Anyways, I hope this has helped slightly in some way, and made things a tiny bit more clear. Let me know if anything didn't quite make sense, and I'd be happy to explain it in further detail, and/or look deeper at the code you are using to perform these low power modes. :)

**Comment by bablokb** [comment]
Thanks, this is very helpful and I really appreciate that you  take the time to explain these details.

> On the bright side, i have a hunch that it might be due to interactions of some other piece of code with the POWMAN system, for example, generated C code / CPU instructions involved in controlling something else on the RP2350 other than POWMAN's low power mode interface itself, and this seperate system is somehow sending events, IRQs, or something else which is causing the CPU to wake up.

Yes, that is exactly what I think too. CircuitPython is a complex program and I am trying to hunt down what is happening. The problem is that "something else" is not showing up in LAST_SWCORE_PWRUP. This register is always the same, regardless of a random wakeup or a wakeup I actually triggered by toggling the wakeup GPIO.

**Comment by dwrrehman** [comment]
One idea which also came to mind, if that indeed may be the case, is to try holding in a "reset" state all peripherals on the RP2350, prior to going to sleep in the low power mode you are using (P1.x). This could in theory be done by writing _almost_ all ones to the `RESETS` register, found in RESETS_BASE in the RP2350's address map, and described in section `7.5.`, titled "Subsystem Resets", and the RESETS register is described on page 506. My idea would be that, hypothetically If you were to write (almost) all ones to this register prior to executing the code that enters low power mode, the problematic peripherals that could be waking up the CPU from sleep could be disabled during sleep mode, and _during the process of entering sleep mode_, which might make a difference. (Note, you wouldn't want to write _all_ ones: bits 10 and 7 would need to be 0 still, as PADS_QSPI, and IO_QSPI subsystems are required to be operational to still talk with the Flash chip, which presumably is where the code is being stored!). 

(As a side note, Another Idea which popped into my head is how the use of an external debugger might affect the program's execution / entering/exiting of low power sleep modes, as if the debugger is not prepared to interface with a CPU that is being power gated, unexpected things could happen, possibly including the debugger force-waking the CPU, if I had to guess. I should importantly note that I have never used an external debugger for doing my RISC-V assembly RP2350 development, nor any debugger at all other than 8 LEDs lol, so I've never had to deal with this intricacy.)

Anyways, however, this is just an idea, obviously diving into the complex inner workings of CircuitPython would be required to truly get to the bottom of why the CPU is waking up here, presumably. I'll give the datasheet a read over again to see if there is any obvious peripheral which could be waking the CPU up upon going into a low power sleep mode like this!

**Comment by bablokb** [comment]
Yes, the external debugger seems to cause problems, see the comment above by @eightycc https://github.com/adafruit/circuitpython/issues/9491#issuecomment-2952711328. I am not using it though. Regarding the other suggestions: I will put them on my list and test them.

**Comment by bablokb** [comment]
I think I solved the problem. I added some delays (mainly to give the system time to flush out trace-messages). CP has two APIs for that: `mp_hal_delay_ms()` and `mp_hal_delay_us()`. The former runs additional code, the latter is just a busy-wait. After switching to the latter I don't see the random behavior anymore.

There is still some way to go to make this usable in the context of CP, but at least the basics are working now.
